packages/itwin/grouping-mapping-widget/src/widget/components/QueryBuilder.ts

```diff
@@ -157,333 +205,259 @@ export class QueryBuilder {
-        }
+
+      if (foundJoin) {
+        foundJoin.value = propertyValue;
+        return;
       }
+
+      foundPropertyClass.classProperties.push(queryJoin);
+      return;
     }
+
+    this.query.unions[unionIndex].properties.push(queryProperty);
   }
 
   public buildQueryString() {
+    this.regenerateQuery();
     if (
       this.query === undefined ||
-      this.query.unions.length === 0 ||
-      this.query.unions[0].classes.length === 0
+      (this.query.unions.find((u) => u.classes.length === 0 && u.properties.length === 0))
     ) {
       return "";
     }
 
     const unionSegments: string[] = [];
     for (const union of this.query.unions) {
-      const querySegments: string[] = [];
 
       const baseClass = union.classes[0];
-      const baseClassName = baseClass.className;
-      const baseIdName = baseClass.isAspect
-        ? `${baseClassName}.Element.id`
-        : `${baseClassName}.ECInstanceId`;
-
-      const selectSegment = `SELECT ${baseIdName}${baseClass.isAspect ? " ECInstanceId" : ""} FROM ${baseClassName}`;
-      querySegments.push(selectSegment);
-
-      if (union.classes.length > 1) {
-        for (let i = 1; i < union.classes.length; i++) {
-          const joinClass = union.classes[i];
-          const joinClassName = joinClass.className;
-          const joinIdName = joinClass.isAspect
-            ? `${joinClassName}.Element.id`
-            : `${joinClassName}.ECInstanceId`;
-
-          const joinSegment = joinClass.isRelational
-            ? ` JOIN ${joinClassName}`
-            : ` JOIN ${joinClassName} ON ${joinIdName} = ${baseIdName}`;
-          querySegments.push(joinSegment);
-
-          for (let j = 0; j < joinClass.properties.length; j++) {
-            const prefix = j === 0 && joinClass.isRelational ? "ON" : "AND";
-            const propertySegment = this.propertySegment(joinClassName, joinClass.properties[j], prefix);
-            querySegments.push(propertySegment);
-          }
-          // when base is regular property, link base to first joined relational property
-          if (joinClass.isRelational && !baseClass.isRelational && i === 1) {
-            querySegments.push(` AND ${joinIdName} = ${baseIdName}`);
-          }
+      const baseProperty = union.properties[0];
+      const baseClassName = baseClass ? baseClass.className : baseProperty?.className;
+      const baseIdName = `${baseClassName}.ECInstanceId`;
+
+      const selectClause = this.selectClause(baseProperty, baseClass);
+
+      let querySegments: Map<string, string[]> = new Map();
+      querySegments = this.relationalJoinSegments(union.classes, querySegments);
+
+      const whereSegments: string[] = [];
+      for (const property of union.properties) {
+        if (property.isCategory) {
+          if (property.className !== baseClassName)
+            querySegments.set("bis.GeometricElement3d", [`bis.GeometricElement3d.ECInstanceId = ${baseIdName}`]);
```

Comment:
How are the keys of this map used?

I worry that there will be duplicate entries for the same class. For example, `bis` is an alias of `BisCore` schema, so `bis.GeometricElement3d` and `BisCore.GeometricElement3d` is the same thing, but would have multiple entries in the `querySegments` map.

Reply:
Yes that would be a problem.
Changed from `bis` to `BisCore`, this should prevent the issue, as query builder uses only class names and not aliases.